### PR TITLE
Fix: ServiceAttribute is missing foregroundServiceType

### PR DESCRIPTION
**Pull request by gugavaro**
Our tools currently do not automatically add new manifest properties.
Manually adding ForegroundServiceType property to ServiceAttribute class

Tracking issue: https://github.com/xamarin/xamarin-android/issues/3712

### CODE REVIEW

File: src/Xamarin.Android.Build.Tasks/Mono.Android/ServiceAttribute.Partial.cs [hunk 1]
```diff
@@ -31,7 +33,16 @@ partial class ServiceAttribute {
 			  "exported",
 			  self          => self.Exported,
 			  (self, value) => self.Exported  = (bool) value
-			}, {
+			},
+			#if ANDROID_29
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Mono.Android/ServiceAttribute.Partial.cs]: This code shouldn't be Conditional.  (It *can*, be, but *shouldn't* be, because if it is removed, things won't work as expected at packaging time, and tracking that down will be annoying.)

File: src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.props.in [hunk 2]
```diff
@@ -14,8 +14,8 @@
 		<AndroidUseAapt2 Condition=" '$(AndroidUseAapt2)' == '' ">True</AndroidUseAapt2>
 		<AndroidPackageNamingPolicy Condition=" '$(AndroidPackageNamingPolicy)' == '' ">LowercaseCrc64</AndroidPackageNamingPolicy>
 		<AndroidUseManagedDesignTimeResourceGenerator Condition=" '$(AndroidUseManagedDesignTimeResourceGenerator)' == '' And '$(OS)' != 'Windows_NT' ">False</AndroidUseManagedDesignTimeResourceGenerator>
-		<AndroidSdkBuildToolsVersion Condition="'$(AndroidSdkBuildToolsVersion)' == ''">28.0.3</AndroidSdkBuildToolsVersion>
-		<AndroidSdkPlatformToolsVersion Condition="'$(AndroidSdkPlatformToolsVersion)' == ''">28.0.0</AndroidSdkPlatformToolsVersion>
+		<AndroidSdkBuildToolsVersion Condition="'$(AndroidSdkBuildToolsVersion)' == ''">29.0.0</AndroidSdkBuildToolsVersion>
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.props.in]: These changes should probably be in a separate PR.  I don't see why they need to be in this one.

**Comment by jonpryor** [comment]
This should also add a test, probably by just using the new attribute in `src/Mono.Android/Test`, but something in `src/Xamarin.Android.Build.Tasks/Tests` may also be desirable.

Regardless, there's a fundamental problem: the [`//service/@android:foregroundServiceType`](https://developer.android.com/guide/topics/manifest/service-element#foregroundservicetype) attribute wants a limited set of strings: 

```
["connectedDevice" | "dataSync" |
    "location" | "mediaPlayback" | "mediaProjection" |
    "phoneCall"]
```

These do map (poorly) with the `ForegroundService` enum members, but nothing actually converts between the two.  Thus, two things need to be done:

1. Write a conversion method between the `ForegroundService` enum and a string:

```csharp
		static string ToString (ForegroundService value)
		{
			switch (value) {
				case ForegroundService.TypeConnectedDevice:   return "connectedDevice";
				...
				default:
					throw new ArgumentException ($"Unsupported ForegroundService value '{value}'.", nameof(value));
			}
		}
```

2. Add this method to the dictionary at:

https://github.com/xamarin/xamarin-android/blob/39bcbb85d662c18e5cdbd4bed091980f68288028/src/Xamarin.Android.Build.Tasks/Utilities/ManifestDocumentElement.cs#L180

```csharp
static readonly Dictionary<Type, Func<object, ICustomAttributeProvider, IAssemblyResolver, int, string>> ValueConverters = new Dictionary<Type, Func<object, ICustomAttributeProvider, IAssemblyResolver, int, string>> () {
    // ...
    { typeof (ForegroundService),     (value, p, r, v) => ToString ((ForegroundService) value) },
};
```

This should ensure that the generated `//service/@android:foregroundServiceType` is one of the expected & correct string values, not...a number, or the enum member name, or whatever the current code will be generating.

File: src/Xamarin.Android.Build.Tasks/Mono.Android/ServiceAttribute.Partial.cs (hunk 1 above)
Comment (gugavaro) [reply src/Xamarin.Android.Build.Tasks/Mono.Android/ServiceAttribute.Partial.cs]: Fixed

File: src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.props.in (hunk 2 above)
Comment (gugavaro) [reply src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.props.in]: Fixed

File: src/Xamarin.Android.Build.Tasks/Utilities/ManifestDocumentElement.cs [hunk 3]
```diff
@@ -365,6 +366,21 @@ static string ToString (string value, ICustomAttributeProvider provider, IAssemb
...
 		}
 
+		static string ToString (ForegroundService value)
+		{
+			switch (value) {
+				case ForegroundService.TypeNone: 		return "none";
+				case ForegroundService.TypeConnectedDevice:	return "connectedDevice";
+				case ForegroundService.TypeDataSync:		return "dataSync";
+				case ForegroundService.TypeLocation:		return "location";
+				case ForegroundService.TypeManifest:		return "manifest";
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Utilities/ManifestDocumentElement.cs]: Oddly, `manifest` isn't a documented value at: https://developer.android.com/guide/topics/manifest/service-element

I'm not sure what to make of that.  Perhaps it shouldn't be supported, and should hit the `default` case?

Comment (gugavaro) [reply src/Xamarin.Android.Build.Tasks/Utilities/ManifestDocumentElement.cs]: Good catch!
I opened Android Studio and it is not an option on the given list. So I guess it is not supported and should hit default case.

Comment (jonpryor) [reply src/Xamarin.Android.Build.Tasks/Utilities/ManifestDocumentElement.cs]: This is something to keep in mind if/when you get around to auto-producing these custom attributes based on the manifest XML: the XML may contain values which are not documented and which Android Studio doesn't "support".

I'm not sure if this can be fully automated.

(Though even partial automation would presumably be an improvement.)

**Comment by jahmai** [comment]
Does this allow for multiple foreground service types?

```
android:foregroundServiceType
Specify that the service is a foreground service that satisfies a particular use case. For example, a foreground service type of "location" indicates that an app is getting the device's current location, usually to continue a user-initiated action related to device location.

*You can assign multiple foreground service types to a particular service.*
```
